src/ucp/rndv/rndv.c

```diff
@@ -1426,6 +1495,162 @@ static void ucp_rndv_do_rkey_ptr(ucp_request_t *rndv_req, ucp_request_t *rreq,
                                       &worker->rkey_ptr_cb_id);
 }
 
+void ucp_rndv_ppln_data_destroy(ucp_context_h context,
+                                ucp_rndv_ppln_rkey_ptr_data_t *ppln_data)
+{
+    ucs_status_t UCS_V_UNUSED status;
+
+    ucp_rkey_destroy(ppln_data->rkey);
+    status = ucp_mem_rereg_mds(context, 0, NULL, 0, 0, NULL,
+                               UCS_MEMORY_TYPE_UNKNOWN, NULL,
+                               &ppln_data->uct_memh, &ppln_data->md_map);
+
+    ucs_assertv(status == UCS_OK, "%s", ucs_status_string(status));
+}
+
+static UCS_F_ALWAYS_INLINE void *
+ucp_rndv_get_frag_rkey_ptr(ucp_worker_h worker, ucp_ep_h ep,
+                           ucp_rndv_rtr_hdr_t *rtr_hdr,
+                           ucs_memory_type_t remote_mem_type,
+                           ucs_memory_type_t local_mem_type,
+                           ucp_rkey_h *rkey_p)
+{
+    ucp_context_h context                    = worker->context;
+    ucp_md_index_t rkey_ptr_md_index         = UCP_NULL_RESOURCE;
+    ucp_rkey_h rkey                          = NULL;
+    ucp_rndv_ppln_rkey_ptr_data_t *ppln_data = NULL;
+    int found                                = 0;
+    ucp_md_map_t rkey_ptr_md_map;
+    ucp_md_index_t md_index;
+    const uct_md_attr_t *md_attr;
+    ucp_ep_h mem_type_ep;
+    ucp_lane_index_t mem_type_rma_lane;
+    unsigned rkey_index;
+    void *local_ptr;
+    khiter_t iter;
+    int ret;
+    ucs_status_t status;
+
+    if ((local_mem_type == UCS_MEMORY_TYPE_HOST) ||
+        (local_mem_type == UCS_MEMORY_TYPE_UNKNOWN)) {
+        return NULL;
+    }
+
+    mem_type_ep = worker->mem_type_ep[local_mem_type];
+    if (mem_type_ep == NULL) {
+        return NULL;
+    }
+
+    ucs_for_each_bit(md_index, ucp_rkey_packed_md_map(rtr_hdr + 1)) {
+        md_attr = &context->tl_mds[md_index].attr;
+        if ((md_attr->cap.flags & UCT_MD_FLAG_RKEY_PTR) &&
+            /* Do not use xpmem, because cuda_copy registration will fail and
+             * performance will not be optimal. */
+            !(md_attr->cap.flags & UCT_MD_FLAG_REG) &&
+            (md_attr->cap.access_mem_types & UCS_BIT(remote_mem_type))) {
+            rkey_ptr_md_index = md_index;
+            break;
+        }
+    }
+
+    if (rkey_ptr_md_index == UCP_NULL_RESOURCE) {
+        return NULL;
+    }
+
+    rkey_ptr_md_map = UCS_BIT(rkey_ptr_md_index);
```

Comment:
seems this var used only once, so not needed

Reply:
removed